### ISSUE TITLE
How can I specify the path of my pem key

### ISSUE THREAD

**Issue by vics80** [lorisleiva/laravel-deployer#12]
My deploy server is an EC2 AWS of Amazon that works only with pem ssh keys. How can I specifi the path of my key?

**Comment by lorisleiva** [comment]
Hi there, I'm not familiar with AWS but you can configure your host to choose a custom private SSH key.

```php
// config/deploy.php

'hosts' => [
    'domain.com' => [
        // ...
        'identityFile' => '~/.ssh/my_private_key',
    ],
],
```

Is that solving your issue?

**Comment by lorisleiva** [comment]
Could you please try to use the `git_tty` option and tell me if you have the same error?

```php
// config/deploy.php

'options' => [
    // ...
    'git_tty' => true,
],
```

**Comment by vics80** [comment]
Now if I run php artisan deploy I get:

> Output:
>   ================
> 
> 
>   Error Output:
>   ================

And If I run de SSH command seems to be permission problem again

**Comment by lorisleiva** [comment]
Okay. I will need more information from you like the content of your `config/deploy.php` and the output of your `php artisan deploy -vvv`.

Also, could you please try to add your host in the `~/.ssh/config` if not already done. Something like that:

```
Host domain.com
  HostName 1.2.3.4
  Port 22
  IdentitiesOnly yes
  IdentityFile ~/.ssh/my_private_key
```

If you then remove the `IdentityFile` option on your `config/deploy.php` file, it will probably use your custom private key for connecting but your normal `id_rsa` key for github.

**Comment by vics80** [comment]
config/deploy.php
`<?php

return [

    /*
    |--------------------------------------------------------------------------
    | Default deployment strategy
    |--------------------------------------------------------------------------
    |
    | This option defines which deployment strategy to use by default on all
    | of your hosts. Laravel Deployer provides some strategies out-of-box
    | for you to choose from explained in detail in the documentation.
    |
    | Supported: 'basic', 'firstdeploy', 'local', 'pull'.
    |
    */

    'default' => 'basic',

    /*
    |--------------------------------------------------------------------------
    | Custom deployment strategies
    |--------------------------------------------------------------------------
    |
    | Here, you can easily set up new custom strategies as a list of tasks.
    | Any key of this array are supported in the `default` option above.
    | Any key matching Laravel Deployer's strategies overrides them.
    |
    */

    'strategies' => [
        //
    ],

    /*
    |--------------------------------------------------------------------------
    | Hooks
    |--------------------------------------------------------------------------
    |
    | Hooks let you customize your deployments conveniently by pushing tasks 
    | into strategic places of your deployment flow. Each of the official
    | strategies invoke hooks in different ways to implement their logic.
    |
    */

    'hooks' => [
        // Right before we start deploying.
        'start' => [
            //
        ],
        
        // Code and composer vendors are ready but nothing is built.
        'build' => [
            //
        ],
        
        // Deployment is done but not live yet (before symlink)
        'ready' => [
            'artisan:storage:link',
            'artisan:view:clear',
            'artisan:cache:clear',
            'artisan:config:cache',
            'artisan:optimize',
            'artisan:migrate',
        ],
        
        // Deployment is done and live
        'done' => [
            //
        ],
        
        // Deployment succeeded.
        'success' => [
            //
        ],
        
        // Deployment failed.
        'fail' => [
            //
        ],
    ],

    /*
    |--------------------------------------------------------------------------
    | Deployment options
    |--------------------------------------------------------------------------
    |
    | Options follow a simple key/value structure and are used within tasks
    | to make them more configurable and reusable. You can use options to
    | configure existing tasks or to use whithin your own custom tasks.
    |
    */

    'options' => [
        'application' => env('APP_NAME', 'Laravel'),
        'repository' => 'git@bitbucket.org:victorvgm/cdpar.git',
        'git_tty' => true,
    ],

    /*
    |--------------------------------------------------------------------------
    | Hosts
    |--------------------------------------------------------------------------
    |
    | Here, you can define any domain or subdomain you want to deploy to.
    | You can provide them with roles and stages to filter them during
    | deployment. Read more about how to configure them in the docs.
    |
    */

    'hosts' => [
        //'ec2-54-229-187-251.eu-west-1.compute.amazonaws.com' => [
        '54.229.187.251' => [
            'deploy_path' => '/var/www/cdpar',
            'user' => 'ubuntu',
            'identityFile' => '~/.ssh/torvic.pem'
        ],
    ],

    /*
    |--------------------------------------------------------------------------
    | Localhost
    |--------------------------------------------------------------------------
    |
    | This localhost option give you the ability to deploy directly on your
    | local machine, without needing any SSH connection. You can use the
    | same configurations used by hosts to configure your localhost.
    |
    */

    'localhost' => [
        //
    ],

    /*
    |--------------------------------------------------------------------------
    | Include additional Deployer recipes
    |--------------------------------------------------------------------------
    |
    | Here, you can add any third party recipes to provide additional tasks, 
    | options and strategies. Therefore, it also allows you to create and
    | include your own recipes to define more complex deployment flows.
    |
    */

    'include' => [
        //
    ],

    /*
    |--------------------------------------------------------------------------
    | Use a custom Deployer file
    |--------------------------------------------------------------------------
    |
    | If you know what you are doing and want to take complete control over
    | Deployer's file, you can provide its path here. Note that, without
    | this configuration file, the root's deployer file will be used.
    |
    */

    'custom_deployer_file' => false,

];`

output of php artisan deploy -vvv:

> • done on [54.229.187.251]
> [localhost] > git rev-parse --abbrev-ref HEAD
> [localhost] < master
> ✈︎ Deploying master on 54.229.187.251 with basic strategy
> • done on [54.229.187.251]
> • done on [54.229.187.251]
> ➤ Executing task deploy:prepare
> [54.229.187.251] > echo $0
> [54.229.187.251] < bash
> [54.229.187.251] > if [ ! -d /var/www/cdpar ]; then mkdir -p /var/www/cdpar; fi
> [54.229.187.251] > if [ ! -L /var/www/cdpar/current ] && [ -d /var/www/cdpar/current ]; then echo 'true'; fi
> [54.229.187.251] > cd /var/www/cdpar && if [ ! -d .dep ]; then mkdir .dep; fi
> [54.229.187.251] > cd /var/www/cdpar && if [ ! -d releases ]; then mkdir releases; fi
> [54.229.187.251] > cd /var/www/cdpar && if [ ! -d shared ]; then mkdir shared; fi
> • done on [54.229.187.251]
> ✔ Ok [1s 436ms]
> ➤ Executing task deploy:lock
> [54.229.187.251] > if [ -f /var/www/cdpar/.dep/deploy.lock ]; then echo 'true'; fi
> [54.229.187.251] > touch /var/www/cdpar/.dep/deploy.lock
> • done on [54.229.187.251]
> ✔ Ok [605ms]
> ➤ Executing task deploy:release
> [54.229.187.251] > cd /var/www/cdpar && (if [ -h release ]; then echo 'true'; fi)
> [54.229.187.251] < true
> [54.229.187.251] > cd /var/www/cdpar && (rm -rf "$(readlink release)")
> [54.229.187.251] > cd /var/www/cdpar && (rm release)
> [54.229.187.251] > cd /var/www/cdpar && (if [ -d releases ] && [ "$(ls -A releases)" ]; then echo 'true'; fi)
> [54.229.187.251] > cd /var/www/cdpar && (if [ -d /var/www/cdpar/releases/1 ]; then echo 'true'; fi)
> [54.229.187.251] > cd /var/www/cdpar && (date +"%Y%m%d%H%M%S")
> [54.229.187.251] < 20180423112818
> [54.229.187.251] > cd /var/www/cdpar && (echo '20180423112818,1' >> .dep/releases)
> [54.229.187.251] > cd /var/www/cdpar && (mkdir /var/www/cdpar/releases/1)
> [54.229.187.251] > cd /var/www/cdpar && (if [[ $(man ln 2>&1 || ln -h 2>&1 || ln --help 2>&1) =~ '--relative' ]]; then echo 'true'; fi)
> [54.229.187.251] < true
> [54.229.187.251] > cd /var/www/cdpar && (ln -nfs --relative /var/www/cdpar/releases/1 /var/www/cdpar/release)
> • done on [54.229.187.251]
> ✔ Ok [2s 621ms]
> ➤ Executing task deploy:update_code
> [54.229.187.251] > if hash command 2>/dev/null; then echo 'true'; fi
> [54.229.187.251] < true
> [54.229.187.251] > command -v 'git'
> [54.229.187.251] < /usr/bin/git
> [54.229.187.251] > /usr/bin/git version
> [54.229.187.251] < git version 1.9.1
> [54.229.187.251] > cd /var/www/cdpar && (if [ -h /var/www/cdpar/release ]; then echo 'true'; fi)
> [54.229.187.251] < true
> [54.229.187.251] > cd /var/www/cdpar && (readlink /var/www/cdpar/release)
> [54.229.187.251] < releases/1
> [54.229.187.251] > cd /var/www/cdpar && (/usr/bin/git clone -b master --depth 1 --recursive -q git@bitbucket.org:victorvgm/cdpar.git /var/www/cdpar/releases/1 2>&1)
> Connection closed by 104.192.143.1
> fatal: Could not read from remote repository.
> 
> Please make sure you have the correct access rights
> and the repository exists.
> Connection to 54.229.187.251 closed.
> ➤ Executing task deploy:failed
> • done on [54.229.187.251]
> ✔ Ok [1ms]
> ➤ Executing task deploy:unlock
> [54.229.187.251] > rm -f /var/www/cdpar/.dep/deploy.lock
> • done on [54.229.187.251]
> ✔ Ok [235ms]
> 
> In Process.php line 222:
> 
>   [Symfony\Component\Process\Exception\ProcessFailedException]
>   The command "ssh -A -tt -i ~/.ssh/torvic.pem ubuntu@54.229.187.251 'cd /var/www/cdpar && (/usr/bin/git clone -b master --depth 1 --recursive -q git@bitbucket.org:victorvgm/cdpar.git /var/www/cdpar/re
>   leases/1 2>&1)'" failed.
> 
>   Exit Code: 128(Invalid exit argument)
> 
>   Working directory: /Users/victorgarciamerediz/Sites/cdpart
> 
>   Output:
>   ================
> 
> 
>   Error Output:
>   ================
> 
> 
> Exception trace:
>  Symfony\Component\Process\Process->mustRun() at /Users/victorgarciamerediz/Sites/cdpart/vendor/deployer/deployer/src/Ssh/Client.php:75
>  Deployer\Ssh\Client->run() at /Users/victorgarciamerediz/Sites/cdpart/vendor/deployer/deployer/src/functions.php:304
>  Deployer\run() at /Users/victorgarciamerediz/Sites/cdpart/vendor/deployer/deployer/recipe/deploy/update_code.php:90
>  Deployer\Deployer::Deployer\{closure}() at n/a:n/a
>  call_user_func() at /Users/victorgarciamerediz/Sites/cdpart/vendor/deployer/deployer/src/Task/Task.php:105
>  Deployer\Task\Task->run() at /Users/victorgarciamerediz/Sites/cdpart/vendor/deployer/deployer/src/Executor/SeriesExecutor.php:63
>  Deployer\Executor\SeriesExecutor->run() at /Users/victorgarciamerediz/Sites/cdpart/vendor/deployer/deployer/src/functions.php:416
>  Deployer\invoke() at /Users/victorgarciamerediz/Sites/cdpart/vendor/lorisleiva/laravel-deployer/src/recipe/laravel-deployer.php:70
>  Deployer\Deployer::Deployer\{closure}() at n/a:n/a
>  call_user_func() at /Users/victorgarciamerediz/Sites/cdpart/vendor/deployer/deployer/src/Task/Task.php:105
>  Deployer\Task\Task->run() at /Users/victorgarciamerediz/Sites/cdpart/vendor/deployer/deployer/src/Executor/SeriesExecutor.php:63
>  Deployer\Executor\SeriesExecutor->run() at /Users/victorgarciamerediz/Sites/cdpart/vendor/deployer/deployer/src/Console/TaskCommand.php:142
>  Deployer\Console\TaskCommand->execute() at /Users/victorgarciamerediz/Sites/cdpart/vendor/symfony/console/Command/Command.php:252
>  Symfony\Component\Console\Command\Command->run() at /Users/victorgarciamerediz/Sites/cdpart/vendor/symfony/console/Application.php:865
>  Symfony\Component\Console\Application->doRunCommand() at /Users/victorgarciamerediz/Sites/cdpart/vendor/deployer/deployer/src/Console/Application.php:133
>  Deployer\Console\Application->doRunCommand() at /Users/victorgarciamerediz/Sites/cdpart/vendor/symfony/console/Application.php:241
>  Symfony\Component\Console\Application->doRun() at /Users/victorgarciamerediz/Sites/cdpart/vendor/symfony/console/Application.php:143
>  Symfony\Component\Console\Application->run() at /Users/victorgarciamerediz/Sites/cdpart/vendor/deployer/deployer/src/Deployer.php:331
>  Deployer\Deployer::run() at /Users/victorgarciamerediz/Sites/cdpart/vendor/deployer/deployer/bin/dep:120
> 
> deploy [-p|--parallel] [-l|--limit LIMIT] [--no-hooks] [--log LOG] [--roles ROLES] [--hosts HOSTS] [-o|--option OPTION] [--] [<stage>]

**Comment by lorisleiva** [comment]
Thanks, and you confirm that you can git clone directly from the server successfully? I.e. your server's SSH key is registered within bitbucket and your server has bitbucket's IP address in its known_hosts?

Did you try editing your `~/.ssh/config` as mentioned on my previous message?

**Comment by vics80** [comment]
I created a new ssh-keygen with another name. And set up this new key in bitbucket. Also I create the ~/.ssh/config file with propper settings for bitbucket and the new key. The config/deploy.php is the same file I sent you.
Now I can clone directly form server, but not from php artisan deploy. The error message is the same.

**Comment by lorisleiva** [comment]
Very strange. When you clone directly from the server, are you connected as `ubuntu` as well?

**Comment by vics80** [comment]
yes

**Comment by vics80** [comment]
In Capistrano, to deploy to a EC2 AWS Server I need to add this in config/deploy/production.rb:

> set :ssh_options, {
>     keys: %w(~/xxx.pem),
>     forward_agent: false,
>     user: fetch(:user),
>     auth_methods: %w(publickey)

The forward_agent is very important, but I don't know where to add this here... maybe in deploy.rb?

**Comment by lorisleiva** [comment]
You can disable the `forward_agent` in your host like so:

```php
// config/deploy.php

'hosts' => [
    'domain.com' => [
        // ...
        'forwardAgent' => false,
    ],
],
```

If you want to play with more ssh options you can have a look at the documentation on [host configurations](https://github.com/lorisleiva/laravel-deployer/blob/master/docs/overview-configure-hosts.md#authentication)

**Comment by vics80** [comment]
Today I set up a fresh instances of EC2, I create a new ssh key (id_rsa) without ~/.ssh/config file and works fine. Here is my config:
```
'hosts' => [
        'xxx.xxx.xxx.xxx' => [
            'deploy_path' => '/var/www/xxx',
            'user' => 'xxx',
            'forwardAgent' => false,
            'identityFile' => '~/.ssh/xxx.pem',
            'sshOptions' => [
                'UserKnownHostsFile' => '/dev/null',
                'StrictHostKeyChecking' => 'no',
                'forwardAgent' => 'false',
],
        ],
```

**Comment by a-bashtannik** [comment]
Same story for deploying using Bitbucket Pipelines and VPS server.

```
Exit Code: 128 (Invalid exit argument)                                                                           
                                                                                                                   
  Host Name: staging.example.com                                                                                     
                                                                                                                   
  ================                                                                                                 
  Cloning into '/var/www/html/example/releases/1'...                                                                   
  Forbidden                                                                                                        
  fatal: Could not read from remote repository.                                                                    
                                                                                                                   
  Please make sure you have the correct access rights                                                              
  and the repository exists.                     

```

adding

```
                'forwardAgent' => false,
```

fixed the issue, thanks @vics80
